### PR TITLE
[TechDraw] Add customizability for dimension line spacing

### LINKED ISSUE #8512: [Feature Request] TechDraw adjust dimension value position
### Is there an existing request for this?

- [X] I have searched the existing requests

### Forums discussion

https://forum.freecad.org/viewtopic.php?t=76122

### Subproject(s) affected?

Techdraw

### Idea description

The spacing between dimension value text and the dimension/extension lines is fixed.  In some cases it would be helpful to have an adjustment to increase this spacing. 

Note to programmer:  see QGIViewDimension::getDefaultIsoDimensionLineSpacing. default is 2*lineWidth.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by benj5378**
Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [x]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

Fixes #8512

### CODE REVIEW

File: src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui [hunk 1]
```diff
@@ -559,6 +559,40 @@ Multiplier of 'Font Size'</string>
...
+          </property>
+          <property name="value">
+           <double></double>
+          </property>
+          <property name="prefEntry" stdset="0">
+           <cstring>LineSpacingISO</cstring>
+          </property>
+          <property name="prefPath" stdset="0">
+           <cstring>Mod/TechDraw/Dimensions</cstring>
+          </property>
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui:593]: I have no idea what these lines define and if they are needed here (copied from GapASME in lines above)

File: src/Mod/TechDraw/Gui/ViewProviderDimension.cpp [hunk 2]
```diff
@@ -86,6 +86,8 @@ ViewProviderDimension::ViewProviderDimension()
                       "Adjusts the gap between dimension point and extension line");
     ADD_PROPERTY_TYPE(GapFactorASME, (Preferences::GapASME()), group, App::Prop_None,
                       "Adjusts the gap between dimension point and extension line");
+    ADD_PROPERTY_TYPE(LineSpacingISO, (-1.0), group, App::Prop_None,
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/ViewProviderDimension.cpp:89]: Why do we need to define the property here too, if it was already defined in the ui file???

File: src/Mod/TechDraw/Gui/QGIViewDimension.cpp [hunk 3]
```diff
@@ -2656,6 +2656,16 @@ double QGIViewDimension::getDefaultIsoDimensionLineSpacing() const
     return Rez::appX(m_lineWidth * 2.0);
 }
 
+double QGIViewDimension::getIsoDimensionLineSpacing() const
+{
+    auto vp = static_cast<ViewProviderDimension*>(getViewProvider(getViewObject()));
+    double customLineSpacing = vp->LineSpacingISO.getValue();
+    if (customLineSpacing == -1.0) {  // By default -1.0 in gui until user defined
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/QGIViewDimension.cpp:2663]: Hardcoded by setting default gui value to -1.0. Any suggestions to how to better check if should use default or not? Using default needs to be an option if it needs to be relative to line width. Maybe a solution could be to set gui value to 2.0 and then returning `m_linewidth * customLineSpacing` in function?

**Comment by benj5378** [comment]
> https://github.com/FreeCAD/FreeCAD/issues/8512#issuecomment-1472967170 : Harder than it sounds as dim text is positioned relative to origin, not the dimension line.
This is still an issue if it is an issue? Also, would preferably be fixed at a later point cause a lot of skill or refactoring is needed first.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9209">pipeline status</a> for feature branch [PR_9209](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9209). Pipeline [830574514](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/830574514) was triggered at [8c7b383](https://github.com/FreeCAD/FreeCAD/pull/9209/commits/8c7b38356c8a48ddab8bbcbc4f81cca8fde4f692). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui (hunk 1 above)
Comment (WandererFan) [reply src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui:593]: There are a bunch of widgets for updating preferences. This is how the "pref" widgets know where to write the parameter value.  Do you have the Designer plugin (https://wiki.freecad.org/Compile_on_Windows/en#Qt_Designer_plugin)?

File: src/Mod/TechDraw/Gui/ViewProviderDimension.cpp (hunk 2 above)
Comment (WandererFan) [reply src/Mod/TechDraw/Gui/ViewProviderDimension.cpp:89]: The Properties in the ViewProvider are persistent,  they are saved with the document.  Properties in the UI apply to the widget.

File: src/Mod/TechDraw/Gui/QGIViewDimension.cpp (hunk 3 above)
Comment (WandererFan) [reply src/Mod/TechDraw/Gui/QGIViewDimension.cpp:2663]: You made a new preference entry above, no?   The ViewProvider ctor would set the property for line spacing to the preference value using a getter from App/Preferences or Gui/PreferencesGui.   Same for initializing the UI - for create, take the value from preferences, for update, take the value from the ViewProvider.

File: src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui (hunk 1 above)
Comment (benj5378) [reply src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui:593]: @WandererFan I have the designer plugin, but I avoided it as it wanted to reformat a lot of stuff in the UI file.

File: src/Mod/TechDraw/Gui/QGIViewDimension.cpp (hunk 3 above)
Comment (benj5378) [reply src/Mod/TechDraw/Gui/QGIViewDimension.cpp:2663]: @WandererFan Not following along? If `-1.0`, it will use default instead.

**Comment by WandererFan** [comment]
> > [#8512 (comment)](https://github.com/FreeCAD/FreeCAD/issues/8512#issuecomment-1472967170) : Harder than it sounds as dim text is positioned relative to origin, not the dimension line.
> 
> This is still an issue - if it is an issue in the first place? Also, would preferably be fixed at a later point cause a lot of skill or refactoring is needed first.

[](https://forum.freecad.org/viewtopic.php?t=77312)

File: src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui (hunk 1 above)
Comment (WandererFan) [reply src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui:593]: The "pref" widgets create/update an entry in the FC parameter file (user.cfg). 

ADD_PROPERTY_TYPE macro adds a persistent property to an object.  The initial value of the property is set by looking up the value in user.cfg.  The look ups are centralized in Preferences/PreferencesGui.

File: src/Mod/TechDraw/Gui/QGIViewDimension.cpp (hunk 3 above)
Comment (WandererFan) [reply src/Mod/TechDraw/Gui/QGIViewDimension.cpp:2663]: You don't decide about using the default here, you just use whatever value is in the ViewProvider's line spacing property.

The ViewProvider property is initialized to the default at creation.

File: src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui (hunk 1 above)
Comment (benj5378) [reply src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui:593]: @WandererFan Right.

`          <property name="prefEntry" stdset="0">
           <cstring>LineSpacingISO</cstring>
          </property>
          <property name="prefPath" stdset="0">
           <cstring>Mod/TechDraw/Dimensions</cstring>
          </property>`
So this automatically sets the preferences. So I just have to add 2.0 to the ui file and we will always have a default preference in user.cfg where LineSpacingISO = 2.0?

Comment (benj5378) [reply src/Mod/TechDraw/Gui/DlgPrefsTechDrawDimensions.ui:593]: @WandererFan

File: src/Mod/TechDraw/Gui/QGIViewDimension.cpp (hunk 3 above)
Comment (benj5378) [reply src/Mod/TechDraw/Gui/QGIViewDimension.cpp:2640]: No. If the value in viewprovider is -1, it uses the default ISO value. And that value can only be calculated from linewidth and not known beforehand

**Comment by benj5378** [comment]
> 1. I'm changing Line Spacing ISO to 1000 and recomputing, but I don't see any change.  What am I missing?
va da fak, let me check, it worked previously.

> 2. is there a reason it is Line Spacing ISO?  is ASME different?
lol, why did I make an ISO function only.

As more customizable properties in ViewProvider is added, it should be made such that there is one dropdown choosing ASME or ISO for all the other properties, instead of having 2 propertyinputs for each property depending on whether it is ISO or ASME.

**Comment by WandererFan** [comment]
https://freecad.github.io/DevelopersHandbook/technical/preferences.html
tell us what is missing in this section.  Better yet, submit a PR to update the handbook.
PR9209_saveRestore

**Comment by WandererFan** [comment]
I don't know why I don't see a difference in these 3 dimensions yet.  Maybe I am not doing something right?  First glance at the code looks reasonable.

PR9209_LineSpacing

**Comment by benj5378** [comment]
Yeah, well, it clearly works on the video? Can you send file plz @WandererFan

**Comment by benj5378** [comment]
user1234:
> for me the moving itself works like in your video in the PR.

**Comment by WandererFan** [comment]
.zip extension to satisfy github...

[PR9209_LineSpacing.FCStd.zip](https://github.com/FreeCAD/FreeCAD/files/11227178/PR9209_LineSpacing.FCStd.zip)

**Comment by WandererFan** [comment]
I had ISO Referencing set as the default style.

- works for ISO Oriented
- doesn't work for ISO Referencing (but should work?)
- doesn't work for ASME Inlined (see below)
- doesn't work for ASME Referencing (see below)

My copy of ASME Y14.5-2009 Dimensioning and Tolerancing doesn't even show
examples of ASME Referencing style, so I think we can ignore the ASME styles until somebody complains.

**Comment by benj5378** [comment]
Yeah, I just figured out. Will look at tomorrow, currently learning Vim so I can work faster and make you sweat even more 😘

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9209">pipeline status</a> for feature branch [PR_9209](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9209). Pipeline [837780607](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/837780607) was triggered at [7ae7408](https://github.com/FreeCAD/FreeCAD/pull/9209/commits/7ae7408434f62f0c88cd07c53ca0b0c38dc37324). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by benj5378** [comment]
@WandererFan So figured out why I never made an ASME line spacing function: ASME dimensions doesn't even have linespacing, likewise there are no logic using any ASME linespacing. ASME dimensions rather has a margin around text than linespacing.

> For the save/restore issue:

```
    ui->pdsbLineSpacingFactor->onSave();
    ui->pdsbLineSpacingFactor->onRestore()
```
Is there a reason that this isn't just done by default from the XML data?

```
double Rez::appX(double x)
{
   return x / getRezFactor();
}
```
What is `RezFactor`?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9209">pipeline status</a> for feature branch [PR_9209](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9209). Pipeline [837798356](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/837798356) was triggered at [97afb02](https://github.com/FreeCAD/FreeCAD/pull/9209/commits/97afb021e8339e397ddf7ae42eb002eb3556022e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by WandererFan** [comment]
"What is RezFactor"

We actually draw the graphics in 0.1 mm units.  If we don't use Rez, the graphics come out like 8 bit game graphics.

**Pull request merged**